### ISSUE TITLE
function instead of manual eval?

### ISSUE THREAD

**Issue by jalbstmeijer** [envwarden/envwarden#1]
Hi,

Thanks for your script, helped me to extend it to something useful for me. 

Wanted to share at least one thing.
Using "eval" manually for each secret env did not seem very practical when not using a separate Docker image for envwarden.

 I choose to define bash functions in my shell to make it easier to use.

example, source this in your .bashrc

function bwset() {
  eval $(envwarden -s "$1")
}

Is this a bad idea?

So when I call 'bwset my-vault-entry", the env vars are sourced immediately.
To be sure it did what expected I made envwarden print the Btiwarden item name selected and env names set.

I implemented several bash functions this way for myself, using an extended version of envwarden.

bwhelp - See this help output.
bwstatus - Show login and vault status.
bwlogin - Login to Bitwarden remote vault, but leave local vault locked.
bwloginu - Login to Bitwarden remote vault and unlock local vault.
bwlogout - Logout from Bitwarden remote vault and remove local vault.
bwunlock - Unlock local vault.
bwlock - Lock local vault.
bwlist - Show local vault's item names.
bwsync - Sync local vault with Bitwarden remote vault (bwlogin already syncs vault).
bwset {string} - Set env vars from vault item matching search string. It looks for text fields starting with ENV_.
bwshow - Show all env vars set by bwset.
bwunset - Unset all env vars set by bwset.

The changes I made are probably very ugly, too ugly to publish. But I'm happy to mail it to you so you can decide to add some parts to your script.

Gr, J

**Comment by gingerlime** [comment]
Hi there @jalbstmeijer and thanks for your input.

I think it's totally fine to create your own aliases and functions and wrap around envwarden. This is part of the Unix philosophy: building small tools that can be combined easily.

Just one note. I'm not sure exactly how you manage your secrets with envwarden, but I get a feeling you define each secret in its own Bitwarden item? It's definitely possible, but I actually think it's easier to create several secrets into one item, and then you can eval all of them together. So you don't need to run `envwarden` lots of times for each secret, but only once for all secrets that you need.

e.g. in this example, envwarden would export 3 secrets (and of course it can do lots more).

```bash
$ envwarden 
.envwarden file not found in /home/xxxxxxxx ... prompting for credentials
? Email address: xxxxxx@example.com
? Master password: [hidden]
export 'DATABASE_PASSWORD'='abcd'
export 'DATABASE_USERNAME'='postgresql'
export 'TOO_GOOD_TO_BE_TRUE'='false'
```

If you want to improve envwarden, that's great. I would appreciate if you can open a pull request so I can review and discuss any changes. If you have any concrete changes in mind, we can also discuss them on this issue or another one. In general, I would prefer to keep envwarden small and simple, using the Unix philosophy in mind :)

I hope you don't mind, but I'll close this issue for now. But feel free to re-open it or create a new issue. Thanks again for reaching out and for sharing your use-case.

**Comment by jalbstmeijer** [comment]
Hi gingertime,

> I think it's totally fine to create your own aliases and functions and wrap around envwarden. This is part of the Unix philosophy: building small tools that can be combined easily.

As I found using eval syntax every time I wanted to have env vars set a bit cumbersome, I wandered why using shell functions was not suggested in the docs.
Tried aliases too, but that seems to not work very well with vars.

So I was wandering if there was a good reason not to suggest it in the docs, like security or something.
From your reply  I conclude it's a useful suggestion to make.

Would such addition to the docs and example functions file to be sourced be welcome as pull request?
 
> Just one note. I'm not sure exactly how you manage your secrets with envwarden, but I get a feeling you define each secret in its own Bitwarden item?

Not necessarily, but I manage lots of separate environments using the same var names, think of AWS api credentials for instance.
I also try not to expose credentials to my shell when they are not needed for a certain task. 

> It's definitely possible, but I actually think it's easier to create several secrets into one item, and then you can eval all of them together. So you don't need to run `envwarden` lots of times for each secret, but only once for all secrets that you need.

Sure, when used together and var names are unique that would be useful.
 
> If you want to improve envwarden, that's great. I would appreciate if you can open a pull request so I can review and discuss any changes.

I'll do.

Issue can be closed.

**Comment by gingerlime** [comment]
> So I was wandering if there was a good reason not to suggest it in the docs, like security or something.
> From your reply I conclude it's a useful suggestion to make.
> Would such addition to the docs and example functions file to be sourced be welcome as pull request?

I'm not sure. I don't expect many tools to recommend how to wrap around them, unless it's a very common use-case. Aliases and shell functions are something that each user can choose to use on their own. So in my opinion, it's not something that should be documented on envwarden itself. Maybe on a wiki page, if at all. Otherwise, it clutters the documentation and makes it hard to see the most important things. If you want to write something short, I can maybe create a wiki page for it. (or not entirely sure if you can suggest a wiki page or create a PR for it? if so, please do). But I'd like to keep the docs focused, so at most would link to this wiki page from the doc.

> Not necessarily, but I manage lots of separate environments using the same var names, think of AWS api credentials for instance.
> I also try not to expose credentials to my shell when they are not needed for a certain task.

In this case, I think a more secure option is to use different Bitwarden users, each user has access only to a specific environment. Otherwise, if you're sharing the same user, then you might end up with the wrong secrets if you just make a simple mistake giving it the wrong search value. Or if someone hacks one of your environments, they can get secrets for other environments as well. If you split things to different Bitwarden users, you're avoiding this issue.

**Comment by jalbstmeijer** [comment]
> I'm not sure. I don't expect many tools to recommend how to wrap around them, unless it's a very common use-case.

True, but this project does that with "eval ${envwarden}", so this was just a suggestion to loose that manual bit. But I understand your point of view if this is a one-time actin un your use case.

>Aliases and shell functions are something that each user can choose to use on their own. So in my opinion, it's not something that should be documented on envwarden itself.

Ok, must have been my shortcoming that it took me a while to realize I could solve it with a  function in by running shell. 

> Maybe on a wiki page, if at all. Otherwise, it clutters the documentation and makes it hard to see the most important things.

I'll consider that.

>> Not necessarily, but I manage lots of separate environments using the same var names, think of AWS api credentials for instance.
> In this case, I think a more secure option is to use different Bitwarden users, each user has access only to a specific environment.

Good point, my way of thinking probably comes form the browser plugin in mind (recently Lastpass), were by it seems natural to have most web site credentials under one vault account

> Otherwise, if you're sharing the same user, then you might end up with the wrong secrets if you just make a simple mistake giving it the wrong search value.

Yes, therefore my extended envwarden version prints the vault item name matched and var names set.
